internal/controller/gateway.go

```diff
@@ -30,47 +32,269 @@ import (
+		}
+	}
+
+	// If needed, finalize the referenced gatewayclass.
+	if !r.isClassFinalized() {
+		// Before doing anything with the gateway, ensure the reference gatewayclass is finalized.
+		if err := r.ensureClassFinalizer(); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to finalize gatewayclass %s: %w",
+				r.referencedClass.Name, err)
+		}
+	}
+
+	// TODO: Ensure the gateway by creating manage infrastructure, i.e. the Envoy service.
+
+	// Pass the gateway off to the eventHandler.
+	r.eventHandler.OnAdd(gw)
 
 	return reconcile.Result{}, nil
 }
+
+// isFinalized returns true if gw is finalized.
+func (r *gatewayReconciler) isFinalized(gw *gatewayapi_v1alpha1.Gateway) bool {
+	for _, f := range gw.Finalizers {
+		if f == gatewayFinalizer {
+			return true
+		}
+	}
+	return false
+}
+
+// ensureFinalizer ensures gw is finalized.
+func (r *gatewayReconciler) ensureFinalizer(gw *gatewayapi_v1alpha1.Gateway) error {
+	if !slice.ContainsString(gw.Finalizers, gatewayFinalizer) {
+		gw.Finalizers = append(gw.Finalizers, gatewayFinalizer)
+		if err := r.client.Update(r.ctx, gw); err != nil {
+			return fmt.Errorf("failed to update %s/%s: %w", gw.Namespace, gw.Name, err)
+		}
+		r.log.WithField("namespace", gw.Namespace).WithField("name", gw.Name).Info("finalized gateway")
+	}
+	return nil
+}
+
+// isClassFinalized returns true if gc is finalized.
+func (r *gatewayReconciler) isClassFinalized() bool {
+	for _, f := range r.referencedClass.Finalizers {
+		if f == gatewayClassFinalizer {
+			return true
+		}
+	}
+	return false
+}
+
+// ensureClassFinalizer ensures a finalizer is applied to the reconciler's referenced gatewayclass.
+func (r *gatewayReconciler) ensureClassFinalizer() error {
+	if !slice.ContainsString(r.referencedClass.Finalizers, gatewayClassFinalizer) {
+		updated := r.referencedClass.DeepCopy()
+		updated.Finalizers = append(updated.Finalizers, gatewayClassFinalizer)
+		if err := r.client.Update(r.ctx, updated); err != nil {
+			return fmt.Errorf("failed to update %s: %w", r.referencedClass.Name, err)
+		}
+		r.log.WithField("name", r.referencedClass.Name).Info("finalized gatewayclass")
+		r.referencedClass = updated
+	}
+	return nil
+}
+
+// ensureGatewayDeleted ensures child resources and finalizers of gw have been deleted.
+func (r *gatewayReconciler) ensureGatewayDeleted(gw *gatewayapi_v1alpha1.Gateway) error {
+	// TODO: Remove managed Envoy infrastructure when introduced.
+
+	// Remove the finalizer from the referenced gateway class, if needed.
+	others, err := r.othersRefClass(gw)
+	if err != nil {
+		return fmt.Errorf("failed to verify if other gateways reference gatewayclass %s: %w",
+			r.referencedClass.Name, err)
+	}
+	if !others {
+		if err := r.EnsureClassFinalizerRemoved(); err != nil {
+			return fmt.Errorf("failed to remove gatewayclass %s finalizer: %w", r.referencedClass.Name, err)
+		}
+	}
+
+	// Remove the finalizer from the gateway.
+	if err := r.EnsureFinalizerRemoved(gw); err != nil {
```

Comment:
I'm not sure if we want to remove the finalizer in this case. It might be safer to keep the finalizer applied so a user is required to investigate further. Thoughts?

Reply:
I don't have strong feelings, either way, just hate when clusters get stuck trying to remove stuff when finalizers are in play. We've had lots of similar trouble with cert-manager in the same way.